aten/src/ATen/native/TensorShape.cpp

```diff
@@ -28,6 +29,136 @@
 #include <vector>
 
 namespace at {
+namespace meta {
+static void cat_check_no_zero_dim(ITensorList tensors) {
+  for (const auto i : c10::irange(tensors.size())) {
+    auto& t = tensors[i];
+    TORCH_CHECK(
+        t.dim() > 0,
+        "zero-dimensional tensor (at position ", i, ") cannot be concatenated");
+  }
+}
+
+static c10::MemoryFormat cat_compute_output_memory_format(ITensorList inputs) {
+  c10::optional<c10::MemoryFormat> format = c10::nullopt;
+  for (auto& t : inputs) {
+    auto f = t.suggest_memory_format();
+    if (f == c10::MemoryFormat::Contiguous) {
+        return f;
+    }
+    if (format.has_value() && format.value() != f) {
+        return c10::MemoryFormat::Contiguous;
+    }
+    format = f;
+  }
+  return format.value();
+}
+
+TORCH_PRECOMPUTE_META_FUNC(cat)(ITensorList tensors, int64_t dim) {
+  // previously, size [0] tensors were the only possible empty tensors; thus, it wasn't possible
+  // to cat empty tensors unless all the other tensors were 1-dimensional, so we allowed these tensors
+  // to be "skipped".  We maintain this behavior for backwards compatibility, but only for this specific
+  // size (i.e. other empty sizes are not skipped).
+  cat_check_no_zero_dim(tensors);
+  dim = at::legacy_cat_wrap_dim(dim, tensors);
+
+  // Checking names before the actual dimensions.
+  auto maybe_outnames = namedinference::compute_cat_outnames(tensors);
+
+  TORCH_CHECK(
+      tensors.size() > 0, "torch.cat(): expected a non-empty list of Tensors");
+
+  // Look for the first valid tensor.
+  size_t valid = tensors.size();
+  for (const auto i : c10::irange(tensors.size())) {
+    if (!at::native::cat_should_skip_tensor(tensors[i])) {
+      valid = i;
+      break;
+    }
+  }
+
+  bool all_contiguous = true;
+  bool all_same_dtype = true;
+  bool all_same_sizes_and_stride = true;
+  auto memory_format = cat_compute_output_memory_format(tensors);
+
+  // Compute what the output dtype should be:
+  const auto& result = maybe_get_output();
+  auto is_out_defined = result.defined();
+  auto out_dtype = at::native::result_type(tensors);
+
+  // If the output tensor is defined, we need to take it into account
+  // when computing the actual output dtype and the flags.
+  if (is_out_defined) {
+    // Check for type promotion, if the output tensor is defined.
+    TORCH_CHECK(
+        canCast(out_dtype, result.scalar_type()),
+        "torch.cat(): input types can't be cast to the desired output type ",
+        result.scalar_type());
+    out_dtype = result.scalar_type();
```

Comment:
Nothing you need to do for this PR, but this makes me wonder if we should also apply the https://github.com/pytorch/pytorch/issues/69813 treatment to dtype promotion/conversion too... I guess it would have to work differently though!

Reply:
Yeah. I think that it would be nice to have that for dtypes, too. I guess, in the end, it would be something similar to what `TensorIterator` kind of already does.